Kubernetes/legos/k8s_check_service_pvc_utilization/k8s_check_service_pvc_utilization.py

```diff
@@ -0,0 +1,207 @@
+        for pvc in pvc_info:
+            print(f"PVC: {pvc['pvc_name']} - Utilized: {pvc['used']} of {pvc['capacity']}")
+        print("-" * 40)
+
+
+def k8s_check_service_pvc_utilization(handle, service_name: str, namespace: str, threshold: int=80) -> Tuple:
+    """
+    k8s_check_service_pvc_utilization checks the utilized disk size of a service's PVC against a given threshold.
+
+    This function fetches the PVC associated with a given service, determines its utilized size,
+    and then compares it to its total capacity. If the used percentage exceeds the provided threshold,
+    it triggers an alert.
+
+    :type handle: object
+    :param handle: Handle object to execute the kubectl command.
+
+    :type service_name: str
+    :param service_name: The name of the service.
+
+    :type threshold: int
+    :param threshold: Percentage threshold for utilized PVC disk size.
+                      E.g., a 80% threshold checks if the utilized space exceeds 80% of the total PVC capacity.
+
+    :type namespace: str
+    :param namespace: The namespace in which the service resides.
+
+    :return: Status and dictionary with PVC name and its size information if the PVC's disk size is below the threshold.
+    """
+    # 1. Get the labels associated with the service
+    get_service_labels_command = f"kubectl get service {service_name} -n {namespace} -o=jsonpath='{{.spec.selector}}'"
+    response = handle.run_native_cmd(get_service_labels_command)
+    if not response or response.stderr:
+        raise ApiException(f"Error while executing command ({get_service_labels_command}): {response.stderr if response else 'empty response'}")
+
+    labels_str = response.stdout.strip().replace("'", "")
+    labels_list = labels_str.split(",")
+    labels_dict = json.loads(response.stdout.replace("'", "\""))  # Convert to valid JSON format
+    label_selector = ",".join([f"{k}={v}" for k, v in labels_dict.items()])
+
+    # 2. Identify a running pod that is utilizing the target PVC using the fetched labels
+    get_pod_command = f"kubectl get pods -n {namespace} -l {label_selector} -o=jsonpath='{{.items[0].metadata.name}}'"
+    response = handle.run_native_cmd(get_pod_command)
+    if not response or response.stderr:
+        raise ApiException(f"Error while executing command ({get_pod_command}): {response.stderr if response else 'empty response'}")
+    pod_name = response.stdout.strip()
+    print(f"Identified pod utilizing the PVC: {pod_name}")
+
+    # 3. Get the PVC name mounted in the identified Pod
+    get_pvc_name_command = f"kubectl get pod {pod_name} -n {namespace} -o=jsonpath='{{.spec.volumes[?(@.persistentVolumeClaim)].persistentVolumeClaim.claimName}}'"
+    response = handle.run_native_cmd(get_pvc_name_command)
+    if not response or response.stderr:
+        raise ApiException(f"Error while executing command ({get_pvc_name_command}): {response.stderr if response else 'empty response'}")
+    pvc_name = response.stdout.strip()
```

Comment:
What if there are more than one pvc associated with a service?

Reply:
Yep. We need to fetch all PVCs in that case. Modified the code to do the same.